test/functorch/test_ops.py

```diff
@@ -450,7 +449,7 @@ def wrapped_fn(*args, **kwargs):
         xfail('nn.functional._scaled_dot_product_attention', device_type='cuda'),
 
         xfail('nn.functional.rrelu'),  # in-place test errors out with no formula implemented
-        xfail('NumpyExpMarkDirtyAutogradFunction'),  # https://github.com/pytorch/pytorch/issues/90225
+        xfail('NumpyExpMarkDirtyAutogradFunction'),  # TODO: calling in-place operation that would mutate a captured Tensor
```

Comment:
This errors for a different reason now, need to investigate.

Reply:
@zou3519 I think I figured out what the issue was with this, but not sure what the solution is

In cpp functorch:
- we create dual tensor with tangent that is captured (so it has a immutable wrapper)
- when we call into exp_, the `checkForInvalidMutationOnCaptures` does not care about the tangent having a immutable wrapper, because that is hidden by the dual tensors wrapper which isn't immutable.
- before we call into VariableType, we first exclude the dynamicLayerFront
- forward AD formula for inplace ops does tangent.copy_(new_tangent). Because we already excluded dynamicLayerFront, we just go into VariableType again (which is basically a noop since neither tensor has tangent this time around)

In pyfunctorch
- we still create that immutable wrapper for tangent
- when we call process, we do not exclude dynamicLayerFront
- process constructs the single layer autograd Function and calls apply (which calls into forward, then jvp)
- after forward is done (with no problems), jvp is performed, which does tangent.mul_(output). At this point, JvpTransform is still at the stop of the stack.
- since we did not exclude this time, we go into dynamicLayerFront this time, which now errors due to checkForInvalidMutationOnCaptures because now we're performing an in-place op on the tangent which has the immutable wrapper.

Reply:
>  jvp is performed, which does tangent.mul_(output)

Where is the mul_ in the code?

Reply:
in the jvp `NumpExp_` defines

Reply:
Maybe one solution could be:
- We could just say that it is okay for us to go through `process` again. It is basicaly a noop, since the stack is the same. Technically it does more checks, but maybe that is fine and we actually want those checks?
- Currently in the creation of a dual tensor, the primal and tangent are not explicitly wrapped, instead we rely on them to get automatically lifted. If we manually wrap tangent (and primal) instead, this error should no longer trigger even if we go through `process` an extra time. Since tangent is something the user passed in themselves, we should be okay with mutating it, and not mark it with the immutable wrapper.

Alternate solution (doesn't work):
- I also tried excluding manually in PyFunctorch's process to mimic the cpp version but ran into an issue with `unwrapped_count > 0 INTERNAL ASSERT FAILED` in the dead tensor wrapper fallback and not sure what that means yet.

Reply:
I'm still processing what is going on, but let me reply to your questions:

> I also tried excluding manually in PyFunctorch's process to mimic the cpp version but ran into an issue with unwrapped_count > 0 INTERNAL ASSERT FAILED in the dead tensor wrapper fallback and not sure what that means yet. (What does this mean?)

There's an invariant that a Tensor with a FuncTorchTensorWrapper dispatch key must be a TensorWrapper. Given that we hit the dead_tensor_fallback, then at least one of the inputs must be a TensorWrapper. The assertion is complaining that none of the inputs are TensorWrapper

Reply:
The thing I am struggling a bit with right now is, does the in-place mutation check even make sense for forward-mode AD?
- if it does, then it sounds like C++ functorch is wrong because it bypasses it
- if it doesn't, then to what extent can we just get rid of it from C++ and Python functorch?

Reply:
Claim: the input-mutation check makes sense for forward-mode AD. We want to prevent a situation where the dual tensor is created on the wrong TensorWrapper.

There are two cases here:

### Case 1: captured value mutated in-place. If we have:
```py
y = torch.tensor(1.)
def f(x):
  y.copy_(x)
  return x + y

jvp(f, (x,), (t,)
```
Then the dual should be created on the wrapped version of y, not y itself. The in-place error checks should ideally raise an error in this situation.

### Case 2: tangent tensor mutated in-place (which is what is happening in this PR). 

```py
import torch
import torch.autograd.forward_ad as fwAD

x = torch.tensor(2.)
y = torch.tensor(3.)

with fwAD.dual_level():
    x_dual = fwAD.make_dual(x, y)
    y.copy_(x_dual)
    x, x_tangent = fwAD.unpack_dual(x_dual)
```
If we ran the functorch.jvp equivalent of the above, it's important that the tangent of x is a TensorWrapper, because it ends up getting its own tangent value.

### Solution?

Given the above, I like one of the solutions you proposed above, which is:

> Currently in the creation of a dual tensor, the primal and tangent are not explicitly wrapped, instead we rely on them to get automatically lifted. If we manually wrap tangent (and primal) instead, this error should no longer trigger even if we go through process an extra time. Since tangent is something the user passed in themselves, we should be okay with mutating it, and not mark it with the immutable wrapper.

functorch.jvp should wrap the primal and the tangent before calling make_dual. The end state is that we get TensorWrapper(primal) that has a tangent which is TensorWrapper(tangent)

Reply:
Case 2 should actually be a bug in PyTorch forward mode AD. Even if we make primal and tangent both have TensorWrapper, the tangent's TensorWrapper should itself never have a tangent. Normally we'd want to error if we're setting a tangent that itself has a tangent, but we're getting around that check with an in-place lol.

That being said, I still think that it is a good idea to manually wrap tangent to indicate that it is a tensor explicitly passed in so that its AD metadata isn't immutable.

I think that morally tangent should not be wrapped. Tangent is being wrapped today because we are computing with it while JVP is active, in theory we are only computing with plain tensors at that point, so (if the forward/backward AD kernels were separate) we should be able exclude Autograd key and properly unwrap and pop JVP off the stack before computing forward grads.